mountpoint-s3/src/data_cache/disk_data_cache.rs

```diff
@@ -0,0 +1,315 @@
+///
+/// TODO: Store additional metadata with each block such as expected S3 key, ETag, etc..
+pub struct DiskDataCache {
+    block_size: u64,
+    cache_directory: PathBuf,
+}
+
+impl DiskDataCache {
+    /// Create a new instance of an [DiskDataCache] with the specified `block_size`.
+    pub fn new(cache_directory: PathBuf, block_size: u64) -> Self {
+        DiskDataCache {
+            block_size,
+            cache_directory,
+        }
+    }
+
+    /// Get the relative path for the given block.
+    fn get_path_for_key(&self, cache_key: &CacheKey) -> PathBuf {
+        let mut path = self.cache_directory.join("v1");
+
+        // An S3 key may be up to 1024 UTF-8 bytes long, which exceeds the maximum UNIX file name length.
+        // Instead, we encode the key and split into 255 character long directory names.
+        let encoded_s3_key = Base64UrlUnpadded::encode_string(cache_key.s3_key.as_bytes());
+        let mut slice = encoded_s3_key.as_str();
+        while !slice.is_empty() {
+            let (chunk, remaining) = slice.split_at(255.min(slice.len()));
+            path.push(chunk);
+            slice = remaining;
+        }
+
+        path.push(cache_key.etag.as_str());
+        path
+    }
+
+    /// Get path for the given block.
+    fn get_path_for_block(&self, cache_key: &CacheKey, block_idx: BlockIndex) -> PathBuf {
+        let mut path = self.get_path_for_key(cache_key);
+        path.push(format!("{}.block", block_idx));
+        path
+    }
+}
+
+impl DataCache for DiskDataCache {
+    fn get_block(&self, cache_key: &CacheKey, block_idx: BlockIndex) -> DataCacheResult<Option<ChecksummedBytes>> {
+        let path = self.get_path_for_block(cache_key, block_idx);
+        let mut file = match fs::File::open(path) {
+            Ok(file) => file,
+            Err(err) if err.kind() == ErrorKind::NotFound => return Ok(None),
+            Err(err) => return Err(err.into()),
+        };
+
+        let bytes = BytesMut::with_capacity(self.block_size as usize);
+        let mut writer = bytes.writer();
+        std::io::copy(&mut file, &mut writer)?;
+        let bytes = writer.into_inner().freeze();
+
+        // TODO: Read checksum from block file
+        let checksum = crc32c::checksum(bytes.as_ref());
+        let bytes = ChecksummedBytes::new(bytes, checksum);
+
+        Ok(Some(bytes))
+    }
+
+    fn put_block(&self, cache_key: CacheKey, block_idx: BlockIndex, bytes: ChecksummedBytes) -> DataCacheResult<()> {
+        let path = self.get_path_for_block(&cache_key, block_idx);
+        trace!(?cache_key, ?path, "new block will be created in data cache");
+        fs::create_dir_all(path.parent().expect("path should include cache key in directory name"))?;
+        let mut file = fs::File::create(path)?;
+        let (bytes, _checksum) = bytes.into_inner().map_err(|_| DataCacheError::InvalidBlockContent)?;
+        // TODO: Store checksum
+        file.write_all(&bytes)?;
```

Comment:
Let's make sure to build a correctly checksum-bracketed on-disk layout here. We need to include the key, etag, block index, and data checksum, and then compute a checksum of all that metadata. We should also encode the serialization version somewhere (e.g. as a magic number at the start of the file).

Reply:
Sounds good! I've been putting something together for the on-disk layout, adding a checksum does sound like something we need. I'll add it to the TODO for that next PR.